iphone/TitaniumKit/TitaniumKit/Sources/API/KrollBridge.m

```diff
@@ -348,47 +348,74 @@ - (void)didStartNewContext:(KrollContext *)kroll
   // Load ti.kernel.js (kroll.js equivalent)
   NSURL *bootstrapURL = [TiHost resourceBasedURL:@"ti.kernel.js" baseURL:NULL];
   NSString *source = [AssetsModule readURL:bootstrapURL];
-
-  JSValue *bootstrapFunc = [objcJSContext evaluateScript:source withSourceURL:bootstrapURL];
-  if (objcJSContext.exception != nil) {
-    evaluationError = YES;
-    [TiExceptionHandler.defaultExceptionHandler reportScriptError:objcJSContext.exception inJSContext:objcJSContext];
-  }
-  [bootstrapFunc callWithArguments:@[ global, [[KrollModule alloc] init] ]];
-  if (objcJSContext.exception != nil) {
+  if (source == nil || source.length == 0) {
+    NSLog(@"[ERROR] Error reading ti.kernel.js, source is nil/empty. Most likely due to failure to decrypt via remote policy and device offline.");
     evaluationError = YES;
-    [TiExceptionHandler.defaultExceptionHandler reportScriptError:objcJSContext.exception inJSContext:objcJSContext];
+    // Don't put up a dialog, as the most likely cause here is remote policy encryption/decryption failing to decrypt due to being offline.
+    // And if we have a dialog up, the security violation dialog won't appear
+  } else {
+    JSValue *bootstrapFunc = [objcJSContext evaluateScript:source withSourceURL:bootstrapURL];
+    if (objcJSContext.exception != nil) {
+      NSLog(@"[ERROR] Error eval'ing ti.kernel.js bootstrap script. Please contact Titanium developers and file a bug report.");
+      evaluationError = YES;
+      [TiExceptionHandler.defaultExceptionHandler reportScriptError:objcJSContext.exception inJSContext:objcJSContext];
+    } else if (bootstrapFunc == nil || [bootstrapFunc isUndefined]) {
+      // it didn't export a bootstrap function! Most likely reason is we hosed the file
+      NSLog(@"[ERROR] Error eval'ing ti.kernel.js, bootstrap function is undefined/nil. Please contact Titanium developers and file a bug report.");
+      evaluationError = YES;
+      [TiExceptionHandler.defaultExceptionHandler reportScriptError:objcJSContext.exception inJSContext:objcJSContext];
+    } else {
+      // ti.kernel.js eval'd OK, so now let's run the exported bootstrap function
```

Comment:
Just a question: Can the "success" case be called earlier, e.g. with reverted if statements? It feels like, since this is the "default case", it should be ran as quick as possible. Right now, for every new context, there are at least 4 new if-statements that all need to access JSCore selectors before completing. I hope this does not influence performance, especially for larger projects

Reply:
regardless of the `if` order, it's still going to be branched code under the hood. I don't think this is a significant concern performance-wise. A "context" is only ever created at app startup, or when a background service launches. It's not like we're creating contexts all over the place. The difference here is that we check if the loaded `ti.kernel.js` source is not nil/empty, and that the exported bootstrap function is not nil or undefined. The exception guards already existed.

Reply:
Thanks for pointing this out!